### ISSUE TITLE
EmailActivity fails when run via background task

### ISSUE THREAD

**Issue by bmingles** [OrchardCMS/Orchard#6033]
In Orchard 1.9.2 EmailActivity no longer seems to work when run after a TimeActivity.

    2015-11-10 23:03:47,308 [7] Orchard.Exceptions.DefaultExceptionPolicy - (null) - An unexpected exception was caught
     (null)
    System.NullReferenceException: Object reference not set to an instance of an object.
       at Orchard.Mvc.ViewEngines.Razor.WebViewPage`1.get_DisplayHelperFactory() in c:\git\mysite\orchard\src\Orchard\Mvc\ViewEngines\Razor\WebViewPage.cs:line 70
       at Orchard.Mvc.ViewEngines.Razor.WebViewPage`1.InitHelpers() in c:\git\mysite\orchard\src\Orchard\Mvc\ViewEngines\Razor\WebViewPage.cs:line 168
       at System.Web.Mvc.RazorView.RenderView(ViewContext viewContext, TextWriter writer, Object instance)
       at System.Web.Mvc.BuildManagerCompiledView.Render(ViewContext viewContext, TextWriter writer)
       at Orchard.DisplayManagement.Descriptors.ShapeTemplateStrategy.ShapeTemplateBindingStrategy.RenderRazorViewToString(String path, DisplayContext context) in c:\git\mysite\orchard\src\Orchard\DisplayManagement\Descriptors\ShapeTemplateStrategy\ShapeTemplateBindingStrategy.cs:line 174
       at Orchard.DisplayManagement.Descriptors.ShapeTemplateStrategy.ShapeTemplateBindingStrategy.Render(ShapeDescriptor shapeDescriptor, DisplayContext displayContext, HarvestShapeInfo harvestShapeInfo, HarvestShapeHit harvestShapeHit) in c:\git\mysite\orchard\src\Orchard\DisplayManagement\Descriptors\ShapeTemplateStrategy\ShapeTemplateBindingStrategy.cs:line 157
       at Orchard.DisplayManagement.Descriptors.ShapeTemplateStrategy.ShapeTemplateBindingStrategy.<>c__DisplayClass26.<>c__DisplayClass28.<Discover>b__15(DisplayContext displayContext) in c:\git\mysite\orchard\src\Orchard\DisplayManagement\Descriptors\ShapeTemplateStrategy\ShapeTemplateBindingStrategy.cs:line 133
       at Orchard.DisplayManagement.Descriptors.ShapeAlterationBuilder.<>c__DisplayClass3.<>c__DisplayClass5.<BoundAs>b__2(DisplayContext displayContext) in c:\git\mysite\orchard\src\Orchard\DisplayManagement\Descriptors\ShapeAlterationBuilder.cs:line 56
       at Orchard.DisplayManagement.Implementation.DefaultDisplayManager.Process(ShapeBinding shapeBinding, IShape shape, DisplayContext context) in c:\git\mysite\orchard\src\Orchard\DisplayManagement\Implementation\DefaultDisplayManager.cs:line 203
       at Orchard.DisplayManagement.Implementation.DefaultDisplayManager.Execute(DisplayContext context) in c:\git\mysite\orchard\src\Orchard\DisplayManagement\Implementation\DefaultDisplayManager.cs:line 99
       at Orchard.DisplayManagement.Implementation.DisplayHelper.ShapeExecute(Object shape) in c:\git\mysite\orchard\src\Orchard\DisplayManagement\Implementation\DisplayHelper.cs:line 86
       at Orchard.DisplayManagement.ShapeDisplay.Display(Object shape) in c:\git\mysite\orchard\src\Orchard\DisplayManagement\ShapeDisplay.cs:line 39
       at Orchard.Email.Services.SmtpMessageChannel.Process(IDictionary`2 parameters)
       at Orchard.Messaging.Services.DefaultMessageService.Send(String type, IDictionary`2 parameters) in c:\git\mysite\orchard\src\Orchard\Messaging\Services\DefaultMessageService.cs:line 20
       at Orchard.Events.DelegateHelper.<>c__DisplayClass2e`3.<BuildAction>b__2d(Object target, Object[] p) in c:\git\mysite\orchard\src\Orchard\Events\DelegateHelper.cs:line 122
       at Orchard.Events.DefaultOrchardEventBus.TryInvokeMethod(IEventHandler eventHandler, Type interfaceType, String messageName, String interfaceName, String methodName, IDictionary`2 arguments, IEnumerable& returnValue) in c:\git\mysite\orchard\src\Orchard\Events\DefaultOrchardEventBus.cs:line 84
       at Orchard.Events.DefaultOrchardEventBus.TryInvoke(IEventHandler eventHandler, String messageName, String interfaceName, String methodName, IDictionary`2 arguments, IEnumerable& returnValue) in c:\git\mysite\orchard\src\Orchard\Events\DefaultOrchardEventBus.cs:line 70
       at Orchard.Events.DefaultOrchardEventBus.TryNotifyHandler(IEventHandler eventHandler, String messageName, String interfaceName, String methodName, IDictionary`2 eventData, IEnumerable& returnValue) in c:\git\mysite\orchard\src\Orchard\Events\DefaultOrchardEventBus.cs:line 53

The error seems to originate in Orchard.Environment.WorkContextExtensions from a null WorkContext in httpContext.Items:

    public WorkContext GetContext(HttpContextBase httpContext) {
        return httpContext.Items[_workContextKey] as WorkContext;
    }

The call that passes the httpContext is in Orchard.WorkContextExtensions:

    public static WorkContext GetWorkContext(this RequestContext requestContext) {

        ...

        return workContextAccessor.GetContext(requestContext.HttpContext);
    }

I propose the following change:

    public static WorkContext GetWorkContext(this RequestContext requestContext) {
        
        ...
        
        if (requestContext.HttpContext != null && requestContext.HttpContext.IsBackgroundContext())
        {
            return workContextAccessor.GetContext();
        }
        else
        {
            return workContextAccessor.GetContext(requestContext.HttpContext);
        }
    }

**Comment by sfmskywalker** [comment]
This may have been fixed by a recent changeset pushed to the **dev** branch. Would you mind to give that a try?

**Comment by bmingles** [comment]
Sure. Do you know offhand which file(s) were the focus of the change?

**Comment by bmingles** [comment]
Nevermind, I see the change to WorkContextAccessor. I'll check it out.

**Comment by TFleury** [comment]
We just need to revert 9b91b1b2609439bb50dfc0321e72e5f39a95022f

**Comment by sfmskywalker** [comment]
Fixed in #6104

**Comment by jtkech** [comment]
@sfmskywalker no time right now to try this good idea, so no closed opinion...

But not sure it's good to create here a `WorkContextScope`(`HttpContextScopeImplementation`) and each time an `HttpContextBase` is resolved...

If I understand, it's a way to have e.g `httpContext.Items[_workContextKey]` not null, so it can solve this issue. Maybe, could have been solved with part of this PR #5887 but only committed in the dev branch.

 Anyway, not all issues described here #6073 are solved, particularly existing code that uses directly the `HttpContextAccessor`. Also used by the `WorkContextAccessor.GetContext()`, so with the current 1.9.x implementation the `GetContext()` without parameter will not return the same object...

To adapt to existing code, that's why in this PR #5966 an `HttpContextBase` is resolved when a `WorkContextScope` is created, carried through the `CallContext LogicalData`, and can be retrieved by the `HttpContextAccessor` that can then be used by the `HttpContextBase` factory to return the same object...

Unfortunately, in this PR we just talked about the baseUrl. I know that @sebastienros is overloaded by Orchard 2. I can talk about my findings on `CallContext LogicalData` in a meeting, let me know...

Best
